runtime/src/bank/upgrade_core_bpf_program.rs

```diff
@@ -0,0 +1,177 @@
+    }
+    // The program data account should _not_ exist
+    let program_data_address = get_program_data_address(&program_address);
+    if bank
+        .get_account_with_fixed_root(&program_data_address)
+        .is_some()
+    {
+        return Err(UpgradeCoreBpfProgramError::ProgramHasDataAccount);
+    }
+    Ok(UpgradeConfig {
+        program_address,
+        program_account,
+    })
+}
+
+fn move_program(bank: &Bank, source: UpgradeConfig, target: UpgradeConfig) {
+    // Burn lamports in the target program account
+    bank.capitalization
+        .fetch_sub(target.program_account.lamports, Relaxed);
+    // Transfer source program account to target program account, clear the
+    // source data account, and update the accounts data size delta
+    let (old_data_size, new_data_size) = (
+        target.program_account.data.len(),
+        source.program_account.data.len(),
+    );
+    bank.store_account(&target.program_address, &source.program_account);
+    bank.store_account(&source.program_address, &AccountSharedData::default());
+    bank.calculate_and_update_accounts_data_size_delta_off_chain(old_data_size, new_data_size);
+}
+
+/// Migrate a native program to BPF using a BPF version of the program,
+/// deployed at some arbitrary address.
+///
+/// This function will move the deployed BPF program in place of the native
+/// program by replacing the account at the native program's address with the
+/// deployed BPF program's account.
+///
+/// This function performs a complete overwrite of the account, including the
+/// owner program. The strict requirements for this swap can be found in each
+/// "check" function.
+// Note: This function is off the hot path until a program migration is due.
+#[allow(dead_code)]
+pub fn migrate_native_program_to_core_bpf(
+    bank: &Bank,
+    target: NativeProgram,
+    source_address: &Pubkey,
+) -> Result<(), UpgradeCoreBpfProgramError> {
+    // Source should be a BPF program owned by the non-upgradeable loader
+    // Target should be a native program owned by the native loader
+    let source = check_program(bank, source_address, &BPF_LOADER_ID)?;
+    let target = check_program(bank, &target.id(), &NATIVE_LOADER_ID)?;
+    move_program(bank, source, target);
+    Ok(())
+}
+
+/// Upgrade a core BPF program using a modified version of the program,
+/// deployed at some arbitrary address.
+///
+/// This function will move the modified BPF program in place of the existing
+/// program by replacing the account at the existing program's address with the
+/// modified program's account.
+///
+/// This function performs a complete overwrite of the account, including the
+/// owner program. The strict requirements for this swap can be found in each
+/// "check" function.
+// Note: This function is off the hot path until a program upgrade is due.
+#[allow(dead_code)]
+pub fn upgrade_core_bpf_program(
+    bank: &Bank,
+    target: NativeProgram,
```

Comment:
Because of this parameter, it might make sense to rename the enum from `NativeProgram` to something like `CoreProgram`. Or we can use a type alias, ie: `type NativeProgram = CoreProgram;`.